### ISSUE TITLE
Background FUSE issue with Cygwin 32 bits

### ISSUE THREAD

**Issue by benrubson** [billziss-gh/winfsp#161]
Hi,

When using Cygwin 32 bits on a 64 bits OS, FUSE FS (EncFS) does not start, sounds like it fails to be sent to background.
If started in foreground (with `-f`, or with `-v` which must implies `-f`), then it works.

Tested on Windows 7 64 bits and 2012 R2 64 bits, with WinFsp 2017.2 and 2018.1 B2.
No issue there using Cygwin 64 bits.

I just made a test using Cygwin 32 bits on Windows 7 32 bits, and it works correctly.

Pretty strange.

Thank you 👍 

Ben

**Comment by benrubson** [comment]
`passthrough-cygfuse` has the same behavior / gives the same results.

**Comment by billziss-gh** [comment]
> Of course there's (at least I think) no reason to use Cygwin32 on a 64 bits OS, but we never know what user could do...

I think some people used to prefer it because it has more packages. I do not know if this is still true today. (I use Cygwin64 myself.)

In any case we should make sure that it works in all cases.

A few questions:

- When the file system dies, does it disappear without trace or does it leave some information behind (e.g. in the Windows event log)?
- Does this happen on a single Cyg32 on Win64 system or have you confirmed on more?
- Does this happen if you link against the WinFsp DLL's directly rather than Cygfuse?

There is nothing obvious in the [`fuse_daemonize`](https://github.com/billziss-gh/winfsp/blob/master/inc/fuse/winfsp_fuse.h#L306) implementation that suggests a problem.

**Comment by billziss-gh** [comment]
This seems to work for me (Cygwin32 on Windows 10 64-bit):

```
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ make cygfuse
gcc passthrough-fuse.c -o passthrough-cygfuse -g -Wall `pkg-config fuse --cflags --libs`
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ ./passthrough-cygfuse.exe . y:
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ pgrep pass
bash: pgrep: command not found
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ ps
      PID    PPID    PGID     WINPID   TTY         UID    STIME COMMAND
     1048       1    1048       1048  ?         197609 09:41:23 /cygdrive/c/Users/billziss/Projects/winfsp/tst/passthrough-fuse/passthrough-cygfuse
     3344     868    3344       4408  cons0     197609 09:41:33 /usr/bin/ps
      868       1     868        868  cons0     197609 09:40:55 /usr/bin/bash
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ ls -l /cygdrive/y
total 236
-rw-r--r-- 1 billziss None    620 Jan 12 14:36 Makefile
-rw-r--r-- 1 billziss None    341 Jan 12 14:36 README.md
drwxr-xr-x 1 billziss None      0 Apr 23 15:30 build
-rwxr-xr-x 1 billziss None 175113 Apr 26 09:41 passthrough-cygfuse.exe
-rw-r--r-- 1 billziss None   9640 Apr  2 23:13 passthrough-fuse.c
-rw-r--r-- 1 billziss None   1313 Jan 12 14:36 passthrough-fuse.sln
-rw-r--r-- 1 billziss None  10382 Jan 12 14:36 passthrough-fuse.vcxproj
-rw-r--r-- 1 billziss None    708 Jan 12 14:36 passthrough-fuse.vcxproj.filters
-rw-r--r-- 1 billziss None  16420 Apr  2 23:13 winposix.c
-rw-r--r-- 1 billziss None   2415 Apr  2 23:13 winposix.h
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ uname -a
CYGWIN_NT-10.0-WOW windows 2.9.0(0.318/5/3) 2017-09-12 10:41 i686 Cygwin
billziss@windows:~/Projects/winfsp/tst/passthrough-fuse [master]$ cmd /c ver

Microsoft Windows [Version 10.0.14393]
```

**Comment by benrubson** [comment]
> I think some people used to prefer it because it has more packages.

I see that it also allows to build a (32 bits) binary which would work on both 32 and 64 bits OS.

> When the file system dies, does it disappear without trace or does it leave some information behind (e.g. in the Windows event log)?

No stacktrace, no error message in Windows event log, and no Cygwin log.

> Does this happen on a single Cyg32 on Win64 system or have you confirmed on more?

I have confirmed on a newly built Windows 7 & 2012 R2 64 bits VMs, with last version of Cygwin32 installed from scratch.

> Does this happen if you link against the WinFsp DLL's directly rather than Cygfuse?

I just tried but can't manage to `make winfsp-fuse` with Cygwin32 on a 64 bits Windows. If I swap DLLs files to make the build to succeed, then the program won't start.

**Comment by benrubson** [comment]
> `CYGWIN_NT-10.0-WOW windows 2.9.0(0.318/5/3) 2017-09-12 10:41 i686 Cygwin`

Mine :
`CYGWIN_NT-6.1-WOW win732 2.10.0(0.325/5/3) 2018-02-02 15:21 i686 Cygwin`
Let me give 2.9 a try.

**Comment by billziss-gh** [comment]
>>Does this happen if you link against the WinFsp DLL's directly rather than Cygfuse?
>
>I just tried but can't manage to make winfsp-fuse with Cygwin32 on a 64 bits Windows. If I swap DLLs files to make the build to succeed, then the program won't start.

Argh, yes. When linking against WinFsp directly the `fuse.pc` file in `/cygdrive/c/Program Files (x86)/WinFsp/lib` references the x64 architecture. You would have to manually change to x86, or specify your own `fuse.pc` in the `pkg-config` command line.

**Comment by benrubson** [comment]
> Let me give 2.9 a try.

It works on the same system using Cygwin 2.9 (same version as yours).
A chance you had it !

**Comment by billziss-gh** [comment]
Wow, so something important changed in Cygwin 32 2.9 -> 2.10 ?!

**Comment by billziss-gh** [comment]
Or perhaps a different compiler, etc.

**Comment by billziss-gh** [comment]
BTW, this may also be a problem with Cygwin `fork`. Unfortunately it is not very reliable for a number of reasons (most of them outside Cygwin's control).

**Comment by benrubson** [comment]
> When linking against WinFsp directly the `fuse.pc` (...) would have to manually change to x86

It helps at build time, but then at runtime :
`passthrough-winfsp-fuse.exe: error while loading shared libraries: winfsp-x86.dll: cannot open shared object file: No such file or directory`

> so something important changed in Cygwin 32 2.9 -> 2.10 ?!

I think so too, something has certainly been broken [after 2.9 release](https://sourceware.org/git/?p=newlib-cygwin.git;a=shortlog).

**Comment by benrubson** [comment]
> Does this happen if you link against the WinFsp DLL's directly rather than Cygfuse?

So ! No !
`passthrough-winfsp-fuse` actually works !

**Comment by billziss-gh** [comment]
One final question: have you tried any of the troubleshooting steps in https://cygwin.com/faq/faq.html#faq.using.fixing-fork-failures

It may take me a couple of days to install 2.10 and try it as I am currently busy with some other work.

**Comment by benrubson** [comment]
> have you tried any of the troubleshooting steps

I'm on this, for sure it's not BLODA related, as I used at least 2 different fresh-installed Windows versions.

One related question to this, when U'll have time, what's the difference between linking to CYGFUSE and  linking directly with the WinFsp DLL ? Is there generally a preferred method ?
As build is done with Cygwin in both cases, Cygwin DLLs will still be required, so I can't really see the difference :) Thx !

**Comment by billziss-gh** [comment]
> what's the difference between linking to CYGFUSE and linking directly with the WinFsp DLL

CYGFUSE is a very simple wrapper around WinFsp-FUSE, that forwards FUSE calls. Its reasons of existence in order of importance:

- The WinFsp DLL provides a FUSE compatible API on both native Windows and Cygwin. There are significant differences between the 2 environments (e.g. different `malloc`, different `errno`'s, etc.) and the provided FUSE API must account for those. For this reason, the DLL exports `fsp_fuse_*` symbols rather than `fuse_*` symbols. The `fsp_fuse_*` symbols accept an extra parameter `env` that "describes" the environment used. The CYGFUSE DLL simplifies this by providing `fuse_*` symbols appropriate for use from Cygwin.

- CYGFUSE knows how to find and load WinFsp. The CYGFUSE DLL is installed in `/usr/bin` and is readily available to FUSE programs.

- CYGFUSE makes it easy to build FUSE programs by using `pkg-config`.

**Comment by benrubson** [comment]
Thank you Bill, clearly understood 👍 

> have you tried any of the troubleshooting steps

So none of the given solution worked :-/

I gave several several different Cygwin version a try :
https://cygwin.com/snapshots/
- 2017-12-09 : works, this is Cygwin 2.9
- 2017-10-18 : works
- 2017-11-14 : works
- 2017-12-01 : does not work
- 2017-12-10 : does not work
- 2018-02-02 : does not work, this is current Cygwin 2.10

So I would say something was broken between [2017-11-14 and 2017-12-01](https://sourceware.org/git/?p=newlib-cygwin.git;a=shortlog;pg=2).

**Comment by billziss-gh** [comment]
Thanks for the great troubleshooting. I am getting quite curious about this one and may look at it sooner rather than later :)

**Comment by billziss-gh** [comment]
CYGFUSE has this ugly hack which may actually be related to the problem:

https://github.com/billziss-gh/winfsp/blob/master/opt/cygfuse/cygfuse.c#L55-L73

**Comment by billziss-gh** [comment]
I am able to reproduce on Win10 with Cygwin32 2.10.0.

**Comment by billziss-gh** [comment]
I created a modified version of CYGFUSE and added some tracing calls and a `sleep` after the `daemon` call. This proved that daemonization actually works and it also allowed me to attach a debugger to the daemonized process.

Soon after attaching the debugger and after the `sleep` finished, I got this lovely dialog box:

![capture](https://user-images.githubusercontent.com/15330948/39339104-bb7e1bec-497b-11e8-9a13-766a1d7605b5.PNG)

The status code `0xC0000028` is `STATUS_BAD_STACK`. My (limited) understanding of this status code, is that it will be thrown when the stack looks bad during Windows Structured Exception unwinding. Obviously it is a pretty fatal condition.

The problem happens within the `StartServiceCtrlDispatcher` API, which is [called](https://github.com/billziss-gh/winfsp/blob/master/src/dll/service.c#L237) by the WinFsp DLL during FUSE initialization (and after daemonization if any). This call allows a WinFsp-FUSE file system to also act as a service (if launched within the Windows Service context). The call will fail when launched from the command line, which is expected and benign.

My theory is that `StartServiceCtrlDispatcher` fails (as it supposed to) and this failure is communicated internally as a Windows Structured Exception. During unwinding Windows examines the stack and finds it wrong for some reason.

This is as much debugging as I have had the time to do so far.

**Comment by billziss-gh** [comment]
The problem appears to be caused by the `-fstack-protector` gcc switch. Modifying the passthrough-fuse `Makefile` to add a `-fno-stack-protector` gcc switch fixes the problem.

This Cygwin commit may be related: [ssp: add APIs for Stack Smashing Protection](https://sourceware.org/git/?p=newlib-cygwin.git;a=commitdiff;h=0a5dfdbd1ba3663a54fa1a7de1a6c4a0a3316a6e)

I am inclined to think that this is a Cygwin problem in the WOW64 environment and not related to WinFsp.

**Comment by billziss-gh** [comment]
> Do you want me to open an issue / ask for support through the Cygwin dev list ?

Good idea. This is likely related to their recent stack smashing and SSP commits, although we do not have a simple repro for them.

**Comment by billziss-gh** [comment]
Ben, thanks. I just subscribed to cygwin-developers to follow that conversation.

**Comment by billziss-gh** [comment]
Ping @benrubson. Just wondering whether you will have any time to investigate this further?

**Comment by benrubson** [comment]
_Sorry for my late answer Bill, I was fully stuck last days...
I'm now back on these subjects.
And thank you very much for tracking this down_ 👍

So, Jony on the ML seems to say there was a change regarding compilation options around `SSP`.
Which would confirm your findings.
And Corinna advises to bisect Cygwin (not sure what this means) or to debug the program using GDB.
I'm not really sure how to properly perform the debugging, as I almost never used GDB.
Do we expect to see the relevant info [setting a breakpoint](https://www.cygwin.com/cygwin-ug-net/gdb.html) to Cygwin's `exception::handle` method ?

**Comment by billziss-gh** [comment]
@benrubson thanks. Did not mean to offload this work to you, just wondered if it was something that you were planning to do :)

Re: bisecting. I would probably start at the suspicious commit and see if a Cygwin that includes it has the problem, while a Cygwin that does not include it does not have the problem.

Another thing to consider is whether the gcc defaults changed between Cygwin 2.9 and 2.10. For example, it might be that gcc in Cygwin 2.9 compiles without `fstack-protector` by default, whereas gcc in Cygwin 2.10 compiles with `fstack-protector` by default.

**Comment by benrubson** [comment]
OK Bill :)

> whether the gcc defaults changed between Cygwin 2.9 and 2.10

You certainly mean between the culprit commit I'll find out bisecting, and the the one just before ?

**Comment by billziss-gh** [comment]
> You certainly mean between the culprit commit I'll find out bisecting, and the the one just before ?

Yes. The process is usually:

- Start with a range of commits A < B, where you know that your test passes at A and fails at B.
- Pick the middle commit M between A and B. Build and try your test. If the test passes the suspect commit is between M and B. If the test fails the suspect commit is between A and M.
- Repeat the process.

In our case we already have an initial suspect commit (the SSP commit). So I suggested trying to short-circuit this long process by trying that commit and its predecessor.

**Comment by billziss-gh** [comment]
The `-fno-stack-protector` option no longer works for me as of Cygwin 2.10.0(0.325/5/3).

```
$ uname -a
CYGWIN_NT-10.0-WOW windows 2.10.0(0.325/5/3) 2018-02-02 15:21 i686 Cygwin
```

I have also tested with an executable built under Cygwin 2.9. It work correctly on Cygwin 2.9, but not on Cygwin 2.10. So this may not be related to the compiler after all.

**Comment by billziss-gh** [comment]
An update on this.

Although I have never found the reason that this problem happens on latest Cygwin, I have successfully worked around it. The workaround was included in the latest WinFsp beta (2018.2 B2), but unfortunately I managed to screw things up at a different place and broke daemonization on **ALL** Cygwin platforms in that release.

A further commit (bef5ba7f3b65eeb34a95561ae9b6939e206ef68e) further fixes daemonization, which should now work on **ALL** Cygwin platforms (at least in my testing). It will be included in the next beta release.

**Comment by benrubson** [comment]
Hi Bill,

> The -fno-stack-protector option no longer works for me as of Cygwin 2.10.0(0.325/5/3).

Were you able to figure out what changed between your previous tests and this failing one ?

> Although I have never found the reason that this problem happens on latest Cygwin...

Well you answer here :)

> I have successfully worked around it.

Really good news ! Many thanks for your work !
So the question, how did you manage to find the workaround ? Only by testing ? Or did you find a clue somewhere ?

So finally, issue was not on Cygwin side ?
Can I post to the mailing list to say issue is now closed ?
Or do you think something may remain there ?

I'll be glad to test the next beta and report 👍 

Thank you again !
